File: ValidParentheses.java
import java.util.Scanner;
import java.util.Stack;

/*
 * in case input: '({)})'
 * would return 'valid' or 'invalid'
 *
 */

public class ValidParentheses {

    public static void main(String[] args) {
//        Scanner sc = new Scanner(System.in);
//        String s = sc.nextLine();
//        String s = "([{}])";
        String s = "{()}[]";
//        String s = "[()]{}{[()()]()}";

//        if (isValid(s)) {
        if (is_valid_arr(s)) {
//        if (is_valid(s)) {
            System.out.print("valid");
        } else {
            System.out.print("invalid");
        }
    }

    public static boolean isValid(String s) {

        Stack<Character> temp = new Stack<>();
        // ' } { '
        for (Character c : s.toCharArray()) {
//			if(c == '(' || c == '{' || c=='[') {
//				temp.push(c);
//			}
//			
//			if(c == ')' && !temp.isEmpty() && temp.peek() == '(') {
//				temp.pop();
//			}else if(c == '}' && !temp.isEmpty() && temp.peek() == '{') {
//				temp.pop();
//			}else if(c == ']' && !temp.isEmpty() && temp.peek() == '[') {
//				temp.pop();
//			}

            if (temp.isEmpty()) {
                temp.push(c);
            } else {
                if (c == ')' && temp.peek() == '(' || c == '(' && temp.peek() == ')')
                    temp.pop();
                else if (c == '}' && temp.peek() == '{' || c == '{' && temp.peek() == '}')
                    temp.pop();
                else if (c == ']' && temp.peek() == '[' || c == '[' && temp.peek() == ']')
                    temp.pop();
            }
        }

        return temp.isEmpty();
    }

    public static boolean is_valid(String s) {

        int length;

        do {
            length = s.length();
            s = s.replace("()", "").replace("{}", "").replace("[]", "");
        } while (length != s.length());

        return s.length() == 0;
    }

    public static boolean is_valid_arr(String s) {
        int i = 0;
        char[] ch = new char[s.length()];

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                ch[++i] = c;
            } else {
                if (i >= 0 && (
                        (ch[i] == '(' && c == ')')
                                || (ch[i] == '{' && c == '}')
                                || (ch[i] == '[' && c == ']')
                )) {
                    i--;
                } else {
                    return false;
                }
            }
        }
        return i <=0;
    }
}

File: ValidParentheses.kt
import data.Node
import java.util.*

fun main() {

      //sorted search list
//      listOf(1, 2, 3, 4, 5, 7, 9).sortedListSearch(9).also {
//            if (it)
//                  print("found")
//            else
//                  print("not found")
//      }

      //"linked list"
//      val node1 = Node(value = 1)
//      val node2 = Node(value = 2)
//      val node3 = Node(value = 3)
//
//      //linking next of n1 to n2
//      node1.next = node2
//      node2.next = node3
//      //would be: node1.value -> node2.value -> node3.value
//
//      print(node1)

      "[(]{)}".validParenthesis().also { println(if (it) "valid" else "tidak valid") }
      "[(]{)".validParenthesis().also { println(if (it) "valid" else "tidak valid") }
}

fun String.validParenthesis(): Boolean {
      var temp = Stack<Char>()

      for (c in this.toCharArray()) {
            if (temp.isEmpty())
                  temp.push(c)
            else {
                  if (c == '(' && temp.peek() == ')' || c == ')' && temp.peek() == '(')
                        temp.pop()
                  else if (c == '[' && temp.peek() == ']' || c == ']' && temp.peek() == '[')
                        temp.pop()
                  else if (c == '{' && temp.peek() == '}' || c == '}' && temp.peek() == '{')
                        temp.pop()

            }
      }

      return temp.empty()
}

fun List<Int>.sortedListSearch(find: Int): Boolean {
      val middle: Int = this.size / 2

      if (find <= this[middle])
            for (index in 0..middle) {
                  if (this[index] == find) {
                        println("found inside")
                        return true
                  }
            }
      else {
            for (index in middle until this.size) {
                  if (this[index] == find) {
                        println("found inside else")
                        return true
                  }
            }
      }
      return false
}